[fix] Fix memory leak caused by incorrect close and destruction

BewareMyPower opened the pull request:
Fixes https://github.com/apache/pulsar-client-cpp/issues/55

### Motivation

1. When a producer or consumer is closed, the reference is still stored
   in `ClientImpl`. If a client kept creating producers or consumers,
   the memory usage would not reduce.
2. When the `HandlerBase::connection_` field is modified, the
   `removeProducer` or `removeConsumer` method is not called. Then these
   producers and consumers will be cached in the connection until the
   connection is closed.
3. The `PartitionedProducerImpl` and `MultiTopicsConsumerImpl` have
   cyclic references, when a `Producer` or `Consumer` instance goes out
   of the scope, the destructors are not called. When I used GDB to
   debug them, I found the reference counts were both greater than 1.

### Modifications

Let's use "handlers" to represent "producers and consumers".

1. In `ClientImpl`, use `SynchronizedHashMap` to store references of
   handlers, as well as the `cleanupXXX` methods to remove a handler.
2. Add `HandlerBase::beforeConnectionChange` method, which is called
   before `connection_` is modified. Disallow the access to
  `connection_` from derived classes.
3. Avoid `shared_from_this()` being passed into callbacks in ASIO
   executors for `PartitionedProducerImpl` and
   `MultiTopicsConsumerImpl`.

This PR also unifies the `shutdown` implementations for handlers and
call `shutdown` in the destructors.
1. Cancel the timers
2. Unregister itself from `ClientImpl` and `ClientConnection`
3. Set the create future with `ResultAlreadyClosed`
4. Set the state to `Closed`

It's called when:
- the destructor is called
- `closeAsync` is completed
- `unsubscribeAsync` is completed with ResultOk

### Verifications

`ShutdownTest` is added to verify the following cases:
- a single topic
- a partitioned topic (multiple topics)
- a partitioned topic with regex subscription

`testClose` verifies `shutdown` when `closeAsync` and `unsubscribeAsync`
are called. `testDestructor` verifies `shutdown` when handlers go out of
the scope and the destructors are called.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)

- [x] `doc-not-needed` 
(Please explain why)

- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

BewareMyPower commented:
Mark it as draft for now because I found a serious problem and will fix it soon.

BewareMyPower commented:
Now all tests passed, PTAL @merlimat @Demogorgon314 @RobertIndie @shibd

shibd commented on lib/ConsumerImpl.cc:
```diff
@@ -990,20 +1002,25 @@ void ConsumerImpl::negativeAcknowledge(const MessageId& messageId) {
-    ConsumerImplPtr ptr = get_shared_this_ptr();
+void ConsumerImpl::closeAsync(ResultCallback originalCallback) {
+    auto callback = [this, originalCallback](Result result) {
+        shutdown();
```
When closing the consumer, we also need to remove the client connection. [cnx->removeConsumer(consumerId_);](https://github.com/apache/pulsar-client-cpp/blob/4a864f20530e230ecee7e4ab09430bd65fb9789c/lib/ConsumerImpl.cc#L1057-L1059)

Also, when invoking `unsubscribe`, don't need to remove the `Connection`, right?

shibd replied:
I want to determine whether the processing behavior of unsubscribe and close is consistent.

BewareMyPower replied:
> I want to determine whether the processing behavior of unsubscribe and close is consistent.

Good question. I'm also confused when I read Java client. It's very similar, the only difference is, when you failed to unsubscribe a client, you should keep the state as Ready, i.e. no cleanup. https://github.com/apache/pulsar/blob/cbf5cf58196b05a35b31385c6a42098c598c9b33/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L422

shibd replied:
Yes, I think that after invoking the unsubscribe method, users can still resubscribe. But this is not the case.

BewareMyPower replied:
> When closing the consumer, we also need to remove the client connection

There is an important point that when `HandlerBase::connection_` is modified, we should unregister itself from the connection. Otherwise, the producer or consumer will always be cached in `ClientConnection`. So this PR addes a `beforeConnectionChange` method, which is always called when `setCnx` or `resetCnx` is called.

The `cnx->removeConsumer` is called in `ConsumerImpl::beforeConnectionChange`. Since `shutdown` calls `resetCnx()`, this method is implicitly called in `shutdown`.

Without this design, users have to remove producer or consumer each time the connection changed, like:

```c++
auto cnx = getCnx().lock();
if (cnx) {
    cnx->removeConsumer(consumerId_);
}
connection_ = newConnection;
```

```c++
auto cnx = getCnx().lock();
if (cnx) {
    cnx->removeConsumer(consumerId_);
}
connection_.reset();
```

BewareMyPower replied:
It's the main cause of #55 because the previous code forgot to remove the producer before `connection_.reset()`. When I debugged, I found `cnx` is always empty in the following code block.

https://github.com/apache/pulsar-client-cpp/blob/7f7653b694996f38cf422d8bef24aa1fdfb21c59/lib/ProducerImpl.cc#L720-L723

I didn't look deeper, but it must be caused by the modification of `connection_` field somewhere else.

shibd replied:
I see. Thanks for your explanation.

RobertIndie replied:
I think the key point of the unsubscribe operation is to delete the subscription in the broker.

RobertIndie replied:
I think the key point of the unsubscribe operation is to delete the subscription in the broker.

BewareMyPower replied:
@RobertIndie Yes, this PR just follows what Java client does.

https://github.com/apache/pulsar/blob/cbf5cf58196b05a35b31385c6a42098c598c9b33/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java#L413-L422

merlimat commented on lib/ConsumerImpl.cc:
```diff
@@ -37,15 +37,14 @@ namespace pulsar {
-ConsumerImpl::ConsumerImpl(const ClientImplPtr client, const std::string& topic,
-                           const std::string& subscriptionName, const ConsumerConfiguration& conf,
-                           bool isPersistent,
+ConsumerImpl::ConsumerImpl(ClientImpl& client, const std::string& topic, const std::string& subscriptionName,
```
Why do we need to change from storing a `weak_ptr` to a reference?

The problem with this is that if you keep using a producer/consumer after the client is already destroyed, you would have a segfault. With the `weak_ptr` we are force to always validating it before using it.

BewareMyPower replied:
> The problem with this is that if you keep using a producer/consumer after the client is already destroyed, you would have a segfault.

Actually not. See `TEST_P(ClientCloseTest, testShutdown)`.

> It seems this PR is included more changes than just the part of using a map to keep track of producers/consumers disposal.

And yes. I think it needs another PR for this refactor. For this PR, I will revert these unnecessary changes soon.

merlimat replied:
> Actually not. See TEST_P(ClientCloseTest, testShutdown).

I think it's probably that most method have already state checks at the beginning. 

(Also I'm seeing many places where are currently doing `client_->lock()` and not checking if we got a valid shared_ptr or an empty one...)

> And yes. I think it needs another PR for this refactor. For this PR, I will revert these unnecessary changes soon.

👍 👍

BewareMyPower replied:
> I think it's probably that most method have already state checks at the beginning.

Yes. If the associated `Client` is expired, the producer or consumer should not be able to do anything except a `ResultAlreadyClosed`.

merlimat reviewed (commented):
It seems this PR is included more changes than just the part of using a map to keep track of producers/consumers disposal. 

In particular, I'm not sure why we need to change from weak_ptr to references.

shibd commented on lib/MultiTopicsConsumerImpl.cc:
```diff
@@ -886,3 +906,8 @@ void MultiTopicsConsumerImpl::messageProcessed(Message& msg) {
+
+void MultiTopicsConsumerImpl::beforeConnectionChange(ClientConnection& cnx) {
+    consumers_.forEachValue(
+        [&cnx](const ConsumerImplPtr& consumer) { consumer->beforeConnectionChange(cnx); });
```
For `MultiTopicConsumerImpl`, It shouldn't be connected, right? 

Should it be more reasonable to change this implementation to throw an exception or add an empty implementation to `ConsumerImplBase`?

BewareMyPower replied:
Good suggestion, adopted.

BewareMyPower commented on lib/ConsumerImpl.cc:
```diff
@@ -242,10 +242,11 @@ void ConsumerImpl::handleCreateConsumer(const ClientConnectionPtr& cnx, Result r
         consumerCreatedPromise_.setValue(get_shared_this_ptr());
     } else {
         if (result == ResultTimeout) {
+            client_.cleanupConsumer(this);
```
You are right, in my latest code (WIP), it's removed.

shibd commented on tests/ClientTest.cc:
```diff
@@ -295,3 +292,90 @@ TEST(ClientTest, testMultiBrokerUrl) {
+        ASSERT_EQ(PulsarFriend::getConsumers(client).size(), 1);
+    }  // client is destructed here and `shutdown()` method will be called
+
+    EXPECT_EQ(ResultAlreadyClosed, producer.send(MessageBuilder().setContent("msg").build()));
```
It feels like this test is wrong. The user does not call any `close` or `shutdown` methods and has references to `producer` and `consumer`. However, it does not work properly.

Maybe we still need `weak_ptr` to prevent the client from being destructed in this scenario.

BewareMyPower replied:
Let's use another PR to discuss about this topic. I'm currently removing code that is related to `weak_ptr` to pure reference. And will push a new change soon.

RobertIndie commented on lib/PartitionedProducerImpl.h:
```diff
@@ -17,16 +17,16 @@
 
+class ClientImpl;
+
 class PartitionedProducerImpl : public ProducerImplBase,
```
Seems we need to implement the `beforeConnectionChange` for the PartitionedProducerImpl.

BewareMyPower replied:
No. Because it doesn't implement `HandlerBase` and has no `connection_` field. The `ClientConnection` instance doesn't hold any `PartitionedProducerImpl` reference

BewareMyPower commented on lib/AckGroupingTrackerDisabled.cc:
```diff
@@ -22,6 +22,7 @@
 #include "HandlerBase.h"
 #include "PulsarApi.pb.h"
 #include <pulsar/MessageId.h>
+#include "LogUtils.h"
```
No. It's just because I changed `std::weak_ptr<ClientImpl>` to `ClientImpl&` in this PR and used forward declaration in the header, so the source file doesn't include "LogUtils.h" (it was included explicitly)

Since these unrelated changes increased the difficulty to review, I'm working on removing them now.

BewareMyPower commented on lib/ConsumerImpl.cc:
```diff
@@ -1018,66 +1035,31 @@ void ConsumerImpl::closeAsync(ResultCallback callback) {
-    failPendingBatchReceiveCallback();
+    auto self = get_shared_this_ptr();
+    cnx->sendRequestWithId(Commands::newCloseConsumer(consumerId_, requestId), requestId)
+        .addListener([this, self, callback](Result result, const ResponseData&) { callback(result); });
```
Yes

RobertIndie replied:
Not sure why this caused a compilation error on my computer, but it's fine in the CI.
```
ConsumerImpl.cc:1047:23: error: lambda capture 'this' is not used [-Werror,-Wunused-lambda-capture]
        .addListener([this, self, callback](Result result, const ResponseData&) { callback(result); });

```

BewareMyPower replied:
Maybe because your compiler is Clang, whose checking rules is more strict than GCC. (You can see the `-Wunused-lambda-capture` option is added)

BewareMyPower commented:
When I add another tests, I found cyclic references for `PartitionedProducerImpl` and `MultiTopicsConsumerImpl` that their destructors won't be called even if `Producer` and `Consumer` go out of scope. I will fix them as well.

BewareMyPower commented:
@merlimat @shibd @RobertIndie Thanks for your reviews. Now I have removed the (`ClientImplWeakPtr` -> `ClientImpl&`) changes and fixed the newly discovered cyclic references (`testDestructor` will fail without the fix), PTAL again.

I changed the milestone to 3.0.0, could you include this PR in the next candidate? @merlimat

merlimat commented on lib/HandlerBase.h:
```diff
@@ -44,11 +44,9 @@ class HandlerBase {
-     * have to check if they can get a shared_ptr out of it or not
-     */
-    ClientConnectionWeakPtr getCnx() const { return connection_; }
+    ClientConnectionWeakPtr getCnx() const;
```
nit: for later, we could return the `shared_ptr` here, already locked inside the method

BewareMyPower replied:
Good suggestion. I will do it later

merlimat reviewed (approved):
👍
